tiledb/sm/array/array.h

```diff
@@ -0,0 +1,198 @@
+
+  /**
+   * Computes an upper bound on the buffer sizes required for a read
+   * query, for the input attributes.
+   *
+   * @param subarray The subarray to focus on. Note that it must have the same
+   *     underlying type as the array domain.
+   * @param attributes The attributes to focus on.
+   * @param buffer_sizes The buffer sizes to be retrieved. This is a map from
+   *     the attribute (or coordinates) name to a pair of sizes (in bytes).
+   *     For fixed-sized attributes, the second size is ignored. For var-sized
+   *     attributes, the first is the offsets size and the second is the
+   *     values size.
+   * @return Status
+   */
+  Status compute_max_buffer_sizes(
+      const void* subarray,
+      const std::vector<std::string>& attributes,
+      std::unordered_map<std::string, std::pair<uint64_t, uint64_t>>*
+          max_buffer_sizes) const;
+
+  /**
+   * Opens the array for reading/writing at the current timestamp.
+   *
+   * @param query_type The mode in which the array is opened.
+   * @return Status
+   */
+  Status open(QueryType query_type);
+
+  /** Closes the array and frees all memory. */
+  Status close();
+
+  /**
+   * Returns the fragment metadata of the array. If the array is not open,
+   * an empty vector is returned.
+   */
+  std::vector<FragmentMetadata*> fragment_metadata() const;
+
+  /**
+   * Returns `true` if the array is empty at the time it is opened.
+   * The funciton returns `false` if the array is not open.
+   */
+  bool is_empty() const;
+
+  /** Returns `true` if the array is open. */
+  bool is_open() const;
+
+  /** Retrieves the array schema. Errors if the array is not open. */
+  Status get_array_schema(ArraySchema** array_schema) const;
+
+  /** Retrieves the query type. Errors if the array is not open. */
+  Status get_query_type(QueryType* qyery_type) const;
+
+  /**
+   * Returns the max buffer size given a fixed-sized attribute and
+   * a subarray. Errors if the array is not open.
+   */
+  Status get_max_buffer_size(
+      const char* attribute, const void* subarray, uint64_t* buffer_size);
+
+  /**
+   * Returns the max buffer size given a var-sized attribute and
+   * a subarray. Errors if the array is not open.
+   */
+  Status get_max_buffer_size(
+      const char* attribute,
+      const void* subarray,
+      uint64_t* buffer_off_size,
+      uint64_t* buffer_val_size);
+
+  /**
+   * Re-opens the array. This effectively updates the "view" of the array,
+   * by loading the fragments potentially written after the last time
+   * the array was opened. Errors if the array is not open.
+   *
+   * @note Applicable only for reads, it errors if the array was opened
+   *     for writes.
+   */
+  Status reopen();
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /** The array URI. */
+  URI array_uri_;
+
+  /** `True` if the array has been opened. */
+  bool is_open_;
```

Comment:
I think this needs to be `std::atomic<bool>` to avoid race conditions

Reply:
In a subsequent PR I was going to address thread safety for all the C API functions (extending #769), which would prevent any race conditions. I will add `std::atomic<bool>` anyway though.